hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockSender.java

```diff
@@ -655,8 +655,11 @@ private int sendPacket(ByteBuffer pkt, int maxChunks, OutputStream out,
           if (ioem.startsWith(EIO_ERROR)) {
             throw new DiskFileCorruptException("A disk IO error occurred", e);
           }
+          String causeMessage = e.getCause() != null ? e.getCause().getMessage() : "";
           if (!ioem.startsWith("Broken pipe")
-              && !ioem.startsWith("Connection reset")) {
+              && !ioem.startsWith("Connection reset")
+              && !causeMessage.startsWith("Broken pipe")
+              && !causeMessage.startsWith("Connection reset")) {
```

Comment:
Was going to hit the merge button, got a doubt here. I think we don't have a null check for ``causeMessage``
we should add one
either do 
```causeMessage = causeMessage == null ? "" : causeMessage;```
before the assertions or change the if like
```
          if (!ioem.startsWith("Broken pipe")
              && !ioem.startsWith("Connection reset")
              && (causeMessage == null
              || (!causeMessage.startsWith("Broken pipe")
              && !causeMessage.startsWith("Connection reset")))) {
```

check whichever way looks better, double check the if condition if you plan to use it, I didn't mess up AND with OR, some brackets or something like that

Reply:
> @fanluoo in case you missed it, can you address this
You're right. Let me improve it

Reply:
You're right. Let me improve it

Reply:
> @ayushtkn  
You're right. I think the first way is better and I will improve it